docs/cli/ilab-model-serve.md

```diff
@@ -0,0 +1,101 @@
+
+Specifically the design of the command, its flags/options, and the how a new serving backend will be tested within the project.
+
+## Design
+
+Since the subject of the `ilab model serve` command is a model, regardless of the format of the model, every command takes in the `--model` flag or uses its default value in the config.
+
+`ilab model serve` has a subcommand for each backend it can serve models with. As of this draft, the two backends `ilab` would serve with are `llama-cpp` and `vllm`.
+
+This would lead to the commands:
+
+- `ilab model serve llama-cpp`
+- `ilab model serve vllm`
+
+The flags for `ilab model serve` are either backend agnostic or specific to a subcommand and are prefixed to indicate that. The intention behind this choice is to ensure there is no mix up of flags between backends.
+
+```console
+ilab model serve
+|
+|_______ --model
+|
+|_______ (other backend agnostic flags)
+|
+|_______ llama-cpp
+|        |
+|        |_______ --llama-cpp-num-gpus
+|        |
+|        |_______ --llama-cpp-max-ctx-size
+|        |
+|        |_______ (other llama-cpp specific args)
+|
+|_______ vllm
+|        |
+|        |_______ --vllm-chat-template
+|        |
+|        |_______ (other popular vllm-specific specific args)
+|        |
+|        |_______ --vllm-args
+```
+
+## Command Examples
+
+### Bare-bones but model specific command
+
+```shell
+ilab model serve --model <PATH> 
+```
+
+- Serves the model at `<PATH>`.
+- If the `<PATH>` is the path for a `.gguf` file then `llama-cpp` is automatically used as the model serving backend.
+- If the path is a directory, `vllm` is automatically used as the backend.
+- The path provided to `vllm` can either be to a Hugging Face repo or a directory with `.safetensors`.
+
+### Bare-bones command
+
+```shell
+ilab model serve
+```
+
+- This command has the same behavior as the one above but the `--model-path` is whatever the default model path is in the config. This is the existing behavior of `ilab serve` today.
+
+### Llama-cpp backend specific commands
+
+```shell
+ilab model serve llama-cpp --model <PATH> --llama-cpp-gpu-layers 4 --llama-cpp-max-ctx-size 1024
+```
+
+- This command serves a model with `llama-cpp`.
+- If the path provided is not to a `.gguf` file, this command would error out.
+- The existing flags to `ilab serve` (besides `--model-path` & `--log-file`) are now specific to the llama-cpp backend and are prefixed with `--llama-cpp-` to indicate that.
+
+### vllm backend specific commands
+
+```shell
+ilab model serve vllm --model <PATH> --vllm-chat-template <PATH> --vllm-args "<MORE VLLM SPECIFIC ARGS>"
+```
+
+- This command serves a model with `vllm`.
+- If the path provided is not a directory, this command would error out.
```

Comment:
Instead of error out, I'd suggest it suggest the correct backend. E.g., 


`You specified vllm but you've provided a gguf file which is not supported by vllm. Would you like to run the gguf file using llamacpp?`

Reply:
I think you're suggesting that the command would steer the user to use the correct backend instead of the command being run again.

I think this is an awesome idea but out of scope for a first PR of introducing a `vllm` backend. I'll be sure to make a follow up issue for this feature.